posts/2023-11-17-tictactoe-tutorial.mdx

```diff
@@ -0,0 +1,665 @@
+      ['4', '5', '6'],
+      ['7', '8', '9'],
+      ['1', '4', '7'],
+      ['2', '5', '8'],
+      ['3', '6', '9'],
+      ['1', '5', '9'],
+      ['3', '5', '7'],
+    ];
+
+    let winner = null;
+    let temp = null;
+    let winningPattern = null;
+
+    winningCombinations.forEach((combination) => {
+      if (winner !== null) {
+        return;
+      }
+      for (let [i, c] of combination.entries()) {
+        if (b[c] === undefined) {
+          temp = null;
+          break;
+        } else {
+          if (temp === null) {
+            temp = b[c];
+            continue;
+          } else {
+            if (temp === b[c]) {
+              if (i === 2) {
+                winner = temp;
+                winningPattern = combination;
+              }
+              continue;
+            }
+          }
+        }
+      }
+    });
+
+    setWinner(winner);
+    setWinningPattern(winningPattern);
+    setOpponentJoined(o);
+    setBoxes(b);
+  }, [storeMessages, filterMessages]);
+```
+
+The game will automatically end if a winning condition is met by one of the users.
+
+### Step 9 : Loading component and rendering the boxes
+
+The loading component will be visible if the node is not started or if an opponent player has not joined the room yet.
+
+```js
+  if (!nodeStart || !player || opponentJoined === null) {
+    return <Loading />;
+  }
+```
+
+Here's the code for `Loading.js` component :
+
+```js
+import React from 'react'
+
+export default function Loading() {
+    return (
+        <div className='h-screen bg-black flex items-center w-full justify-center'>
+            <svg width="60" height="60" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+                <g fill="none" stroke="#ffffff" stroke-linecap="round" stroke-width="2">
+                    <path stroke-dasharray="60" stroke-dashoffset="60" stroke-opacity=".3" d="M12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12C3 7.02944 7.02944 3 12 3Z">
+                        <animate fill="freeze" attributeName="stroke-dashoffset" dur="1.3s" values="60;0" />
+                    </path>
+                    <path stroke-dasharray="15" stroke-dashoffset="15" d="M12 3C16.9706 3 21 7.02944 21 12">
+                        <animate fill="freeze" attributeName="stroke-dashoffset" dur="0.3s" values="15;0" />
+                        <animateTransform attributeName="transform" dur="1.5s" repeatCount="indefinite" type="rotate" values="0 12 12;360 12 12" />
+                    </path>
+                </g>
+            </svg>
+            
+        </div>
+    )
+}
+
+```
+
+After you are done setting up the loading component, it's time to write a function that will handle rendering the boxes. This function will be used as an onClick handler that will send either `x` or `o` based on the player who clicks the box. It will also render the boxes with `x` or `o` based on the player who has clicked the box previously. Here's the function :
```

Comment:
```suggestion
After you are done setting up the loading component, it's time to write a function that will handle rendering the boxes. This function will be used as an `onClick` handler that will send either `x` or `o` based on the player who clicks the box. It will also render the boxes with `x` or `o` based on the player who has clicked the box previously. Here's the function :
```